client/network/src/light_client_requests/sender.rs

```diff
@@ -0,0 +1,1330 @@
+						}
+						RequestFailure::Obsolete => {
+							debug_assert!(
+								false,
+								"Can not receive `RequestFailure::Obsolete` after dropping the \
+								 response receiver.",
+							);
+							self.pending_requests.push_back(sent_request.into_pending());
+						}
+						RequestFailure::Network(OutboundFailure::Timeout) => {
+							self.remove_peer(sent_request.peer);
+							self.peerset.report_peer(
+								sent_request.peer,
+								rep::TIMEOUT,
+							);
+							self.pending_requests.push_back(sent_request.into_pending());
+						},
+						RequestFailure::Network(OutboundFailure::UnsupportedProtocols) => {
+							self.remove_peer(sent_request.peer);
+							self.peerset.report_peer(
+								sent_request.peer,
+								ReputationChange::new_fatal(
+									"peer does not support light client or block request protocol",
+								),
+							);
+							self.pending_requests.push_back(sent_request.into_pending());
+						}
+						RequestFailure::Network(OutboundFailure::DialFailure) => {
+							self.remove_peer(sent_request.peer);
+							self.peerset.report_peer(
+								sent_request.peer,
+								ReputationChange::new_fatal(
+									"failed to dial peer",
+								),
+							);
+							self.pending_requests.push_back(sent_request.into_pending());
+						}
+						RequestFailure::Network(OutboundFailure::ConnectionClosed) => {
+							self.remove_peer(sent_request.peer);
+							self.peerset.report_peer(
+								sent_request.peer,
+								ReputationChange::new_fatal(
+									"connection to peer closed",
+								),
+							);
+							self.pending_requests.push_back(sent_request.into_pending());
+						}
+					}
+
+					continue;
+				}
+			};
+
+			match self.on_response(sent_request.peer, &sent_request.request, response) {
+				Ok(reply) => sent_request.request.return_reply(Ok(reply)),
+				Err(Error::UnexpectedResponse) => {
+					log::debug!("Unexpected response from peer {}.", sent_request.peer);
+					self.remove_peer(sent_request.peer);
+					self.peerset.report_peer(
+						sent_request.peer,
+						ReputationChange::new_fatal(
+							"unexpected response from peer",
+						),
+					);
+					self.pending_requests.push_back(sent_request.into_pending());
+				}
+				Err(other) => {
+					log::debug!("error handling response from peer {}: {}", sent_request.peer, other);
+					self.remove_peer(sent_request.peer);
+					self.peerset.report_peer(
+						sent_request.peer,
+						ReputationChange::new_fatal(
+							"invalid response from peer",
+						),
+					);
+					self.pending_requests.push_back(sent_request.into_pending())
+				}
+			}
+		}
+
+		// If we have a pending request to send, try to find an available peer and send it.
+		while let Some(mut pending_request) = self.pending_requests.pop_front() {
+			if pending_request.attempts_left == 0 {
+				pending_request.request.return_reply(Err(ClientError::RemoteFetchFailed));
+				continue
+			}
+
+			pending_request.attempts_left -= 1;
```

Comment:
I know that it is something that has already been in the master and you've just moved that code, but this actually causing this error in logs:
```
Light: [2021-01-21 14:42:48] 2021-01-21 14:42:48.245  DEBUG tokio-runtime-worker sc_network::light_client_requests::sender: No peer available to send request to.    
Light: [2021-01-21 14:42:48] 2021-01-21 14:42:48.246   WARN tokio-runtime-worker txpool: Failed to fetch block body: RemoteFetchFailed   
```

Couple of lines below (see `// Break in case there is no idle peer.`), the request is returned to the queue if there are no peers or all peers are busy with other requests. The initial idea was that request only should fail if we have received enough "reject" from full nodes. E.g. 2 different full nodes have responded with error to the same request.

Probably this is causing an issues? I'd just move this line (`pending_request.attempts_left -= 1;`) just before converting it `into_sent` request. Wdyt?

Reply:
Suggestion sounds good and indeed this solves the remaining connection issues. Thanks @svyatonik!

Commit https://github.com/paritytech/substrate/pull/7895/commits/2477f3474f4baacb6a11682a6a3f7b3ed33bea17 does the corresponding change.